File: trunk/java/signals/org.psem2m.signals.directory/src/org/psem2m/signals/directory/impl/DirectoryUpdater.java
/**
 * File:   DirectoryUpdater.java
 * Author: Thomas Calmant
 * Date:   18 juin 2012
 */
package org.psem2m.signals.directory.impl;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import org.apache.felix.ipojo.annotations.Component;
import org.apache.felix.ipojo.annotations.Invalidate;
import org.apache.felix.ipojo.annotations.Requires;
import org.apache.felix.ipojo.annotations.Validate;
import org.psem2m.isolates.base.IIsolateLoggerSvc;
import org.psem2m.isolates.constants.IPlatformProperties;
import org.psem2m.isolates.constants.ISignalsConstants;
import org.psem2m.isolates.services.dirs.IPlatformDirsSvc;
import org.psem2m.signals.HostAccess;
import org.psem2m.signals.ISignalBroadcaster;
import org.psem2m.signals.ISignalBroadcaster.ESendMode;
import org.psem2m.signals.ISignalData;
import org.psem2m.signals.ISignalDirectory;
import org.psem2m.signals.ISignalDirectory.EBaseGroup;
import org.psem2m.signals.ISignalDirectoryConstants;
import org.psem2m.signals.ISignalListener;
import org.psem2m.signals.ISignalReceiver;
import org.psem2m.signals.ISignalSendResult;
import org.psem2m.signals.IWaitingSignal;
import org.psem2m.signals.IWaitingSignalListener;

/**
 * Java implementation of the directory updater
 * 
 * @author Thomas Calmant
 */
@Component(name = "psem2m-signals-directory-updater-factory")
public class DirectoryUpdater implements ISignalListener,
        IWaitingSignalListener {

    /** The signals directory */
    @Requires
    private ISignalDirectory pDirectory;

    /** The logger */
    @Requires
    private IIsolateLoggerSvc pLogger;

    /**
     * The port to access the directory dumper. The member shall not be changed
     * after component validation
     */
    // private int pDumperPort;

    /** The platform information service */
    @Requires
    private IPlatformDirsSvc pPlatform;

    /** The signal receiver (it must be accessible) */
    @Requires(filter = "(" + ISignalReceiver.PROPERTY_ONLINE + "=true)")
    private ISignalReceiver pReceiver;

    /** The signal sender (it must be usable) */
    @Requires(filter = "(" + ISignalBroadcaster.PROPERTY_ONLINE + "=true)")
    private ISignalBroadcaster pSender;

    /**
     * Retrieves the directory of a remote isolate.
     * 
     * This method is called after a CONTACT signal has been received from a
     * monitor.
     * 
     * @param aSignalData
     *            The received contact signal
     */
    private void grabRemoteDirectory(final ISignalData aSignalData) {

        // Only monitors can send us this kind of signal
        final String remoteId = aSignalData.getSenderUID();
        final String remoteName = aSignalData.getSenderName();
        if (!remoteName
                .startsWith(IPlatformProperties.SPECIAL_ISOLATE_ID_MONITOR)) {
            // Log & Ignore
            pLogger.logWarn(this, "grabRemoteDirectory",
                    "Contacts must be made by monitors only, sender=", remoteId);
            return;
        }

        // Get information on the sender
        final String remoteAddress = aSignalData.getSenderAddress();
        final String remoteNode = aSignalData.getSenderNode();
        final Object rawContent = aSignalData.getSignalContent();
        if (!(rawContent instanceof Map)) {
            pLogger.logWarn(this, "grabRemoteDirectory", "Unreadable content=",
                    rawContent);
            return;
        }

        // Get the dumper port
        final Map<?, ?> content = (Map<?, ?>) rawContent;
        final Integer remotePort = (Integer) content.get("port");
        if (remotePort == null) {
            pLogger.logWarn(this, "grabRemoteDirectory", "No access port given");
            return;
        }

        // Store the node information
        pDirectory.setNodeAddress(remoteNode, remoteAddress);

        // Grab the directory
        try {
            final Object[] results = pSender.sendTo(
                    ISignalDirectoryConstants.SIGNAL_DUMP, null,
                    new HostAccess(remoteAddress, remotePort));

            // Store the dumped directory
            handleDumpedDirectory(results, remoteNode);

        } catch (final Exception e) {
            pLogger.logSevere(this, "grabRemoteDirectory",
                    "Error grabbing the directory of host=", remoteAddress,
                    "port=", remotePort, "id=", remoteId, ":", e);
        }
    }

    /**
     * Handles an ACK signal
     * 
     * @param aSenderId
     *            ID of the signal broadcaster
     */
    private void handleAck(final String aSenderId) {

        pLogger.logDebug(this, "handleSynAck", "ACK from", aSenderId);

        // Our acknowledgment has been received
        if (pDirectory.validateIsolatePresence(aSenderId)) {
            pDirectory.notifyIsolatePresence(aSenderId);
        }
    }

    /**
     * Tries to find the directory dump in the results then stores it
     * 
     * If aIgnoredNode is not null, the address corresponding to it in the
     * dumped directory won't be stored.
     * 
     * @param aResults
     *            The signal results
     * @param aIgnoredNode
     *            The address for this node must be ignored
     */
    private synchronized void handleDumpedDirectory(final Object[] aResults,
            final String aIgnoredNode) {

        if (aResults == null || aResults.length == 0) {
            // No result...
            pLogger.logWarn(this, "grabDirectory",
                    "Nothing returned by the directory dumper");
            return;
        }

        // Get the first map result only
        if (aResults.length != 1) {
            pLogger.logWarn(this, "grabDirectory",
                    "More than one result found. Ignoring others.");
        }

        // Type conversion...
        Map<?, ?> dump = null;
        for (final Object result : aResults) {
            if (result instanceof Map) {
                dump = (Map<?, ?>) result;
                break;
            }
        }

        if (dump != null) {
            // All good, store the dump in the directory
            pDirectory.storeDump(dump, Arrays.asList(aIgnoredNode), null);

            // Send our registration signal to all isolates we known,
            // without propagation
            sendRegistrationToAll(true);

        } else {
            // Nothing found
            pLogger.logWarn(this, "grabDirectory",
                    "No readable result in the dump");
        }
    }

    /*
     * (non-Javadoc)
     * 
     * @see
     * org.psem2m.signals.ISignalListener#handleReceivedSignal(java.lang.String,
     * org.psem2m.signals.ISignalData)
     */
    @Override
    public synchronized Object handleReceivedSignal(final String aSignalName,
            final ISignalData aSignalData) {

        // The sender ID
        final String senderId = aSignalData.getSenderUID();

        if (ISignalDirectoryConstants.SIGNAL_DUMP.equals(aSignalName)) {
            pLogger.logDebug(this, "handleReceivedSignal", "DUMP from",
                    senderId);

            // Register the incoming isolate
            registerIsolate(aSignalData);

            // Dump the directory
            return pDirectory.dump();

        } else if (ISignalDirectoryConstants.SIGNAL_REGISTER
                .equals(aSignalName)) {
            // Register the isolate
            registerIsolate(aSignalData);

        } else if (ISignalDirectoryConstants.SIGNAL_REGISTER_SYNACK
                .equals(aSignalName)) {
            // We received a registration acknowledgment
            handleSynAck(senderId);

        } else if (ISignalDirectoryConstants.SIGNAL_REGISTER_ACK
                .equals(aSignalName)) {
            // Final acknowledgment
            handleAck(senderId);

        } else if (ISignalsConstants.ISOLATE_LOST_SIGNAL.equals(aSignalName)) {
            // Isolate lost
            final String lostIsolate = (String) aSignalData.getSignalContent();
            pDirectory.unregisterIsolate(lostIsolate);

        } else if (ISignalDirectoryConstants.SIGNAL_CONTACT.equals(aSignalName)) {
            // Monitor contact. Should only happen for forkers
            grabRemoteDirectory(aSignalData);
        }

        // No result
        return null;
    }

    /**
     * Handles a SYN-ACK signal
     * 
     * @param aSenderId
     *            ID of the signal broadcaster
     */
    private void handleSynAck(final String aSenderId) {

        pLogger.logDebug(this, "handleSynAck", "SYN-ACK from", aSenderId);

        if (!pDirectory.isRegistered(aSenderId)) {
            // Unknown isolate, ask for a REGISTER signal
            pLogger.logWarn(this, "handleSynAck", "Unknown isolate ID:",
                    aSenderId);

        } else if (pDirectory.synchronizingIsolatePresence(aSenderId)) {

            // Validate the registration (the isolate talked to us)
            pDirectory.validateIsolatePresence(aSenderId);

            // Our registration has been accepted, final acknowledgment
            // (blocking call)
            final ISignalSendResult result = pSender.send(
                    ISignalDirectoryConstants.SIGNAL_REGISTER_ACK, null,
                    aSenderId);

            if (result != null) {
                // Notify our listeners (final step)
                pDirectory.notifyIsolatePresence(aSenderId);

            } else {
                // Failed to talk to the isolate ?
                pLogger.logWarn(this, "handleSynAck", "Isolate ID=", aSenderId,
                        "did not respond to the ACK signal.");
            }

        } else {
            pLogger.logWarn(this, "handleSynAck",
                    "SYN-ACK ignored due to failed state change for isolate=",
                    aSenderId);
        }
    }

    /**
     * Component invalidated
     */
    @Invalidate
    public void invalidate() {

        // Unregister signals
        pReceiver.unregisterListener(
                ISignalDirectoryConstants.SIGNAL_PREFIX_MATCH_ALL, this);

        pReceiver.unregisterListener(ISignalsConstants.ISOLATE_LOST_SIGNAL,
                this);

        pLogger.logInfo(this, "invalidate", "Directory Updater Gone");
    }

    /**
     * Prepares the registration signal content.
     * 
     * @param aPropagate
     *            If true, the receivers of this signal will re-emit it
     * @return The content for a registration signal
     */
    private Map<String, Object> prepareRegistrationContent(
            final boolean aPropagate) {

        // Prepare content
        final Map<String, Object> content = new HashMap<String, Object>();
        content.put(IRegistrationKeys.UID, pPlatform.getIsolateUID());
        content.put(IRegistrationKeys.NAME, pPlatform.getIsolateName());

        // Let the receiver compute our address
        content.put(IRegistrationKeys.ADDRESS, null);
        content.put(IRegistrationKeys.NODE, pPlatform.getIsolateNode());
        content.put(IRegistrationKeys.PORT, pReceiver.getAccessInfo().getPort());
        content.put(IRegistrationKeys.PROPAGATE, aPropagate);

        return content;
    }

    /**
     * Handles an isolate registration signal: registers the isolate in the
     * local directory and propagates the registration if necessary
     * 
     * @param aSignalData
     *            The signal data
     */
    private synchronized void registerIsolate(final ISignalData aSignalData) {

        // Get the ID of the sender
        final String senderUID = aSignalData.getSenderUID();

        // Extract information
        @SuppressWarnings("unchecked")
        final Map<String, Object> content = (Map<String, Object>) aSignalData
                .getSignalContent();

        final String isolateUID = (String) content.get(IRegistrationKeys.UID);
        if (pPlatform.getIsolateUID().equals(isolateUID)) {
            // Ignore self-registration
            return;
        }

        final String name = (String) content.get(IRegistrationKeys.NAME);
        final String node = (String) content.get(IRegistrationKeys.NODE);
        final Integer port = (Integer) content.get(IRegistrationKeys.PORT);
        final Boolean propagate = (Boolean) content
                .get(IRegistrationKeys.PROPAGATE);

        String address;
        if (aSignalData.getSenderNode().equals(node)) {
            /*
             * If both the registered and the registrar are on the same node,
             * use the sender address to update the node access
             */
            address = aSignalData.getSenderAddress();

        } else {
            /*
             * Else use the address indicated in the signal, or use the sender
             * address
             */
            address = (String) content.get(IRegistrationKeys.ADDRESS);
            if (address == null || address.isEmpty()) {
                address = aSignalData.getSenderAddress();
            }
        }

        // 1. Update the node address
        pDirectory.setNodeAddress(node, address);

        // 2. Register the isolate
        final boolean newlyRegistered = pDirectory.registerIsolate(isolateUID,
                name, node, port);

        if (pDirectory.isRegistered(isolateUID)) {
            // 2b. Acknowledge the registration

            if (isolateUID.equals(senderUID)) {
                /*
                 * Case 1: we got the registration from the isolate itself ->
                 * Send a SYN-ACK
                 */
                if (newlyRegistered) {
                    // Update registration state
                    pDirectory.synchronizingIsolatePresence(isolateUID);
                }

                pSender.fire(ISignalDirectoryConstants.SIGNAL_REGISTER_SYNACK,
                        null, isolateUID);

            } else if (newlyRegistered) {
                /*
                 * Case 2: we got the registration by propagation -> Send a
                 * REGISTER to the registered isolate
                 */
                pLogger.logDebug(this, "registerIsolate",
                        "Sending a registration signal to", isolateUID,
                        "due to a propagation");
                pSender.fire(ISignalDirectoryConstants.SIGNAL_REGISTER,
                        prepareRegistrationContent(false), isolateUID);
            }
        }

        // 3. Propagate the registration if needed
        if (propagate.booleanValue()) {
            // Stop propagation
            content.put(IRegistrationKeys.PROPAGATE, Boolean.FALSE);

            // Store the address we used
            content.put(IRegistrationKeys.ADDRESS, address);

            pSender.fireGroup(ISignalDirectoryConstants.SIGNAL_REGISTER,
                    content, EBaseGroup.STORED, isolateUID);
        }
    }

    /**
     * Sends the registration signal to all isolates stored in the directory
     * 
     * @param aPropagate
     *            If true, the receivers of this signal will re-emit it
     */
    private void sendRegistrationToAll(final boolean aPropagate) {

        // Get the content
        final Map<String, Object> content = prepareRegistrationContent(aPropagate);

        // Send the signal (don't wait for a result)
        final ISignalSendResult res = pSender.sendGroup(
                ISignalDirectoryConstants.SIGNAL_REGISTER, content,
                EBaseGroup.STORED);

        if (res == null) {
            // No answer, registration may have failed
            pLogger.logWarn(this, "sendRegistrationToAll",
                    "Registration signal not sent/received");

        } else {
            // Log the receivers
            pLogger.logDebug(this, "sendRegistrationToAll",
                    "Registration sent to=", res.getResults().keySet());
        }
    }

    /**
     * Component validated
     */
    @Validate
    public void validate() {

        pLogger.logDebug(this, "validate", "Directory Updater starting...");

        // Register to signals
        pLogger.logDebug(this, "validate", "Registering to signals...");
        pReceiver.registerListener(
                ISignalDirectoryConstants.SIGNAL_PREFIX_MATCH_ALL, this);
        pReceiver.registerListener(ISignalsConstants.ISOLATE_LOST_SIGNAL, this);

        // Compute the dump port
        final String dumpPortStr = System
                .getProperty(ISignalDirectoryConstants.PROP_DUMPER_PORT);
        int dumperPort = -1;
        try {
            dumperPort = Integer.parseInt(dumpPortStr);

        } catch (final NumberFormatException e) {
            // Ignore
        }

        if (dumperPort <= 0) {
            // Bad port
            pLogger.logWarn(this, "validate", "Unreadable dumper port=",
                    dumpPortStr);

        } else {
            // Retrieve the directory
            pLogger.logDebug(this, "validate", "Grabbing directory from port=",
                    dumperPort);

            // Prepare signal content, without propagating the signal
            final Map<String, Object> content = prepareRegistrationContent(false);

            // Stack the signal
            pSender.stackTo(ISignalDirectoryConstants.SIGNAL_DUMP, content,
                    this, ESendMode.SEND, Integer.MAX_VALUE, new HostAccess(
                            "localhost", dumperPort));
        }

        pLogger.logInfo(this, "validate", "Directory Updater Ready");
    }

    /*
     * (non-Javadoc)
     * 
     * @see
     * org.psem2m.signals.IWaitingSignalListener#waitingSignalSent(org.psem2m
     * .signals.IWaitingSignal)
     */
    @Override
    public void waitingSignalSent(final IWaitingSignal aSignal) {

        if (ISignalDirectoryConstants.SIGNAL_DUMP.equals(aSignal.getName())) {
            // Found
            pLogger.logDebug(this, "waitingSignalSent",
                    "Dump directory signal response received");

            // Call the handling code
            handleDumpedDirectory(aSignal.getSendToResult(), null);

        } else {
            // Ignored result
            pLogger.logDebug(this, "waitingSignalSent", "Signal name=",
                    aSignal.getName(), "sent.");
        }
    }

    /*
     * (non-Javadoc)
     * 
     * @see
     * org.psem2m.signals.IWaitingSignalListener#waitingSignalTimeout(org.psem2m
     * .signals.IWaitingSignal)
     */
    @Override
    public void waitingSignalTimeout(final IWaitingSignal aSignal) {

        pLogger.logWarn(this, "waitingSignalTimeout", "Signal name=",
                aSignal.getName(), "timed out");
    }
}

File: trunk/python/cohorte.python/cohorte/signals/directory_updater.py
#!/usr/bin/env python
# -- Content-Encoding: UTF-8 --
"""
Python implementation of the PSEM2M Directory updater

Created on 18 juin 2012

:author: Thomas Calmant
:license: GPLv3
"""

# Documentation strings format
__docformat__ = "restructuredtext en"

# Boot module version
__version__ = "1.0.0"

# ------------------------------------------------------------------------------

# COHORTE constants
import cohorte

# Pelix/iPOPO
from pelix.ipopo.decorators import ComponentFactory, Requires, Validate, \
    Invalidate

# ------------------------------------------------------------------------------

import logging
import threading

_logger = logging.getLogger(__name__)

# ------------------------------------------------------------------------------

SIGNAL_PREFIX = "/psem2m-directory-updater"
""" Prefix of all directory updater signals """

SIGNAL_PREFIX_MATCH_ALL = "%s/*" % SIGNAL_PREFIX
""" Pattern to match all directory updater signals """

SIGNAL_DUMP = "%s/dump" % SIGNAL_PREFIX
""" Directory dump request """

SIGNAL_REGISTER = "%s/register" % SIGNAL_PREFIX
""" Isolate registration notification """

SIGNAL_REGISTER_SYNACK = "%s/syn-ack" % SIGNAL_REGISTER
""" Isolate registration acknowledgment synchronization """

SIGNAL_REGISTER_ACK = "%s/ack" % SIGNAL_REGISTER
""" Isolate registration acknowledgment """

SIGNAL_CONTACT = "%s/contact" % SIGNAL_PREFIX
""" Special case for early starting forkers: a monitor signals its dump port """

SIGNAL_ISOLATE_LOST = "/psem2m/isolate/lost"
""" Isolate disappeared """

# ------------------------------------------------------------------------------

@ComponentFactory("cohorte-signals-directory-updater-factory")
@Requires("_directory", cohorte.SERVICE_SIGNALS_DIRECTORY)
@Requires("_receiver", cohorte.SERVICE_SIGNALS_RECEIVER)
@Requires("_sender", cohorte.SERVICE_SIGNALS_SENDER)
class DirectoryUpdater(object):
    """
    Directory update component
    """
    def __init__(self):
        """
        Constructor
        """
        self._directory = None
        self._receiver = None
        self._sender = None

        # Lock to have a clean directory during signals handling
        self._lock = threading.Lock()


    def _grab_directory(self, host, port, ignored_node=None):
        """
        Sends a directory dump signal to the dumper on local host.
        
        If ignored_host is not None, the address corresponding to it in the
        dumped directory won't be stored.
        
        :param host: Directory dumper host address
        :param port: Directory signal listener port
        :param ignored_node: The address for this node must be ignored
        """
        # Prepare pre-registration content, without propagation
        content = self._prepare_registration_content(False)

        # Send the dump signal, with a registration content
        # -> we give 5 attempts (~15 seconds) for the dumper to reply
        max_attempts = 5
        attempts = 0
        while attempts < max_attempts:
            attempts += 1
            sig_results = self._sender.send_to(SIGNAL_DUMP, content, host, port)
            if not sig_results or not sig_results["results"]:
                # No result: the dumper might be busy
                _logger.warning("Nothing returned by the directory dumper "
                                "(%d/%d)", attempts, max_attempts)

            else:
                # Got it
                _logger.warning("Grabbed the directory after %d attempts",
                                attempts)
                break

        # Local information
        local_id = self._directory.get_isolate_uid()
        local_node = self._directory.get_local_node()

        # Get the first result only
        results = sig_results["results"]
        if len(results) > 1:
            _logger.warning("More than one result found. Ignoring others")

        result = results[0]

        # 1. Filter the nodes
        ignored_nodes = (ignored_node, local_node)

        # 2. Filter the IDs
        ignored_ids = (local_id,)

        # 3. Call the directory, to do all the update at once
        self._directory.store_dump(result, ignored_nodes, ignored_ids)

        # 4. Now, we can send our registration signal
        self._send_registration_to_all(True)


    def _grab_remote_directory(self, signal_data):
        """
        Retrieves the directory of a remote isolate.
        
        This method is called after a CONTACT signal has been received from a
        monitor.
        
        :param signal_data: The received contact signal
        """
        # Only monitors can send us contacts
        remote_id = signal_data["senderUID"]
        if not remote_id.startswith("org.psem2m.internals.isolates.monitor"):
            _logger.warning("Contacts must be made by a monitor, not %s",
                            remote_id)
            return

        # Get information on the sender
        remote_address = signal_data["senderAddress"]
        remote_node = signal_data["senderNode"]

        _logger.debug("Grab directory from %s (%s)", remote_id, remote_node)

        # Get the dumper port
        content = signal_data["signalContent"]
        remote_port = content["port"]
        if not remote_port:
            _logger.warning("No port given")
            return

        # Store the remote node
        self._directory.set_node_address(remote_node, remote_address)

        # Grab the directory
        self._grab_directory(remote_address, remote_port, remote_node)


    def _prepare_registration_content(self, propagate):
        """
        Prepares the registration signal content.
        
        :param propagate: If true, the receivers of this signal will re-emit it
        :return: The content for a registration signal
        """
        # Beat confirmation
        isolate_id = self._directory.get_isolate_uid()

        # FIXME: setup groups from the configuration service
        groups = ["ALL"]

        if isolate_id.startswith("org.psem2m.internals.isolates.forker"):
            # Forkers can only be forkers
            groups.append("FORKERS")

        else:
            # A forker can't be an isolate and vice versa
            groups.append("ISOLATES")

        if isolate_id.startswith("org.psem2m.internals.isolates.monitor"):
            # A monitor is a special isolate
            groups.append("MONITORS")

        return {"id": isolate_id,
                "address": None,  # <- No address when sending
                "node": self._directory.get_local_node(),
                "port": self._receiver.get_access_info()[1],
                "groups": groups,
                "propagate": propagate}


    def _register_isolate(self, signal_data):
        """
        Registers an isolate according to the given map
        
        :param signal_data: The received signal
        :return: True if the isolate has been registered 
        """
        sender_id = signal_data["senderUID"]
        content = signal_data["signalContent"]

        isolate_uid = content["uid"]
        if isolate_uid == self._directory.get_isolate_uid():
            # Ignore self-registration
            return False

        node = content["node"]
        if node == signal_data["senderNode"]:
            # If both the registered and the registrar are on the same node,
            # use the sender address to update the node access
            address = signal_data["senderAddress"]

        else:
            # Else: use the address indicated in the signal, or use the sender
            # address
            address = content.get("address", None)
            if not address:
                # Address could be empty, so don't use the dict.get() parameter
                address = signal_data["senderAddress"]

        # 1. Update the node host
        self._directory.set_node_address(node, address)

        # 2. Register the isolate
        registered = self._directory.register_isolate(isolate_uid,
                                                      content["name"],
                                                      node,
                                                      content["port"])

        # 2b. Acknowledge the registration, even if we knew it before
        if self._directory.is_registered(isolate_uid):
            if sender_id == isolate_uid:
                # Case 1: we got the registration from the isolate itself
                # -> Send a SYN-ACK
                self._sender.fire(SIGNAL_REGISTER_SYNACK, None,
                                  isolate=isolate_uid)

            elif registered:
                # Case 2: we got the registration by propagation
                # -> Send a REGISTER
                self._sender.fire(SIGNAL_REGISTER,
                                  self._prepare_registration_content(False),
                                  isolate=isolate_uid)

        else:
            _logger.debug("NOT REGISTERED: %s", isolate_uid)

        # 3. Propagate the registration, if needed
        if content["propagate"]:
            # Propagate only once...
            content["propagate"] = False

            # Indicate the address we used for the registration
            content["address"] = address

            self._sender.fire(SIGNAL_REGISTER, content, dir_group="OTHERS",
                              excluded=[isolate_uid])

        return registered


    def _send_registration_to_all(self, propagate):
        """
        Sends the registration signal to all known isolates
        
        :param propagate: If true, the receivers of this signal will re-emit it
        """
        content = self._prepare_registration_content(propagate)

        # Send the registration signal
        results = self._sender.send(SIGNAL_REGISTER, content,
                                    dir_group="OTHERS")

        if not results:
            _logger.warning("Registration signal not sent/received")

        else:
            _logger.debug("Registration sent to: %s",
                          results[0].keys())


    def handle_received_signal(self, name, signal_data):
        """
        Called when a remote services signal is received
        
        :param name: Signal name
        :param signal_data: Signal content
        """
        sender_id = signal_data["senderUID"]

        if name == SIGNAL_DUMP:
            with self._lock:
                # Register the incoming isolate
                _logger.debug("DUMP from %s", sender_id)
                self._register_isolate(signal_data)

                # Dump the directory
                return self._directory.dump()

        elif name == SIGNAL_REGISTER:
            with self._lock:
                # Isolate registration
                self._register_isolate(signal_data)

        elif name == SIGNAL_REGISTER_SYNACK:
            with self._lock:
                # Send the final acknowledgment
                self._sender.fire(SIGNAL_REGISTER_ACK, None, isolate=sender_id)

                # Notify listeners
                self._directory.validate_isolate_presence(sender_id)

        elif name == SIGNAL_REGISTER_ACK:
            with self._lock:
                # Our acknowledgment has been received
                self._directory.validate_isolate_presence(sender_id)

        elif name == SIGNAL_ISOLATE_LOST:
            with self._lock:
                # Unregister the isolate
                lost_isolate = signal_data["signalContent"]
                if lost_isolate:
                    # Unregister it
                    self._directory.unregister_isolate(lost_isolate)

        elif name == SIGNAL_CONTACT:
            # A contact has been signal, ask for a remote directory dump
            # -> Forker only
            with self._lock:
                self._grab_remote_directory(signal_data)


    @Invalidate
    def invalidate(self, context):
        """
        Component invalidated
        """
        # Unregister to isolate registration signals
        self._receiver.unregister_listener(SIGNAL_PREFIX_MATCH_ALL, self)
        self._receiver.unregister_listener(SIGNAL_ISOLATE_LOST, self)


    @Validate
    def validate(self, context):
        """
        Component validate
        """
        # Register to isolate registration signals
        self._receiver.register_listener(SIGNAL_PREFIX_MATCH_ALL, self)
        self._receiver.register_listener(SIGNAL_ISOLATE_LOST, self)

        # Get the local dumper port
        dump_port = context.get_property("psem2m.directory.dumper.port")
        if not dump_port:
            _logger.warning("No local dumper port found.")
            # Can't grab...

        else:
            # Grab from local host
            self._grab_directory("localhost", int(dump_port))
